test/steps/task.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020 Red Hat, Inc. and/or its affiliates
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package steps
+
+import (
+	"fmt"
+
+	"github.com/cucumber/godog"
+	"github.com/cucumber/godog/gherkin"
+	"github.com/kiegroup/kogito-cloud-operator/test/framework"
+)
+
+// registerTaskSteps register all task steps
+func registerTaskSteps(s *godog.Suite, data *Data) {
+	s.Step(`^Service "([^"]*)" contains (\d+) (?:task|tasks) of process with name "([^"]*)" and task name "([^"]*)"$`, data.serviceContainsTasksOfProcessWithNameAndTaskName)
+	s.Step(`^Complete "([^"]*)" task on service "([^"]*)" and process with name "([^"]*)" with body:$`, data.completeTaskOnServiceAndProcessWithName)
+}
+
+func (data *Data) serviceContainsTasksOfProcessWithNameAndTaskName(serviceName string, numberOfTasks int, processName, taskName string) error {
+	routeURI, err := framework.WaitAndRetrieveRouteURI(data.Namespace, serviceName)
+	if err != nil {
+		return err
+	}
+
+	// We assume that one process instance is available, can be refactored if more flexibility needed
+	foundProcessInstances, err := framework.GetProcessInstances(data.Namespace, routeURI, processName)
+	if err != nil {
+		return err
+	}
+	if len(foundProcessInstances) == 0 {
+		return fmt.Errorf("no process instance found, expected one instance")
+	}
+	if len(foundProcessInstances) > 1 {
+		return fmt.Errorf("too many process instances found, expected one instance, but found %d instances", len(foundProcessInstances))
+	}
+
+	// We assume that only tasks with specified name exist, can be refactored if more flexibility needed
+	processInstanceID := foundProcessInstances[0]["id"].(string)
+	foundTasks, err := framework.GetTasks(data.Namespace, routeURI, processName, processInstanceID)
+	if err != nil {
+		return err
+	}
+	for _, foundTaskName := range foundTasks {
+		if taskName != foundTaskName {
+			return fmt.Errorf("found unexpected task name %s", foundTaskName)
+		}
+	}
+	if len(foundTasks) < numberOfTasks {
```

Comment:
could be done before the loop ?

Reply:
Well, I prefer it this way.
Let me show an example - what will happen if we expect one task of name AAA, but task request return two tasks - AAA and BBB.
In case we would check number of tasks first then we get message saying that there are two tasks instead of one. But it doesn't tell me much about why there are two tasks. Is the expected task duplicated? Or some other tasks is created?
In current implementation you get an error saying that unexpected task named BBB was found, which AFAIK provides more information than previous approach.

Ideally we should compare expected and actual result and construct complex error showing what do we expect and what we got, but I wanted to keep things simple.

Reply:
good point. you win :)